src/argilla/server/search_engine/base.py

```diff
@@ -0,0 +1,207 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import dataclasses
+import re
+from abc import ABCMeta, abstractmethod
+from contextlib import asynccontextmanager
+from typing import Any, AsyncGenerator, Dict, Iterable, List, Optional, Union
+from uuid import UUID
+
+from pydantic import BaseModel, Field
+
+from argilla.server.enums import ResponseStatus, ResponseStatusFilter
+from argilla.server.models import Dataset, MetadataProperty, Record, Response, User
+
+__all__ = [
+    "SearchEngine",
+    "UserResponse",
+    "StringQuery",
+    "MetadataFilter",
+    "TermsMetadataFilter",
+    "IntegerMetadataFilter",
+    "FloatMetadataFilter",
+    "UserResponseStatusFilter",
+    "SearchResponseItem",
+    "SearchResponses",
+]
+
+
+class UserResponse(BaseModel):
+    values: Optional[Dict[str, Any]]
+    status: ResponseStatus
+
+
+@dataclasses.dataclass
+class StringQuery:
+    q: str
+    field: Optional[str] = None
+
+
+@dataclasses.dataclass
+class UserResponseStatusFilter:
+    user: User
+    statuses: List[ResponseStatusFilter]
+
+
+@dataclasses.dataclass
+class MetadataFilter:
+    metadata_property: MetadataProperty
+
+    @classmethod
+    @abstractmethod
+    def from_string(cls, metadata_property: MetadataProperty, string: str) -> "MetadataFilter":
+        pass
+
+
+@dataclasses.dataclass
+class TermsMetadataFilter(MetadataFilter):
+    values: List[str]
+
+    @classmethod
+    def from_string(cls, metadata_property: MetadataProperty, string: str) -> "MetadataFilter":
+        return cls(metadata_property=metadata_property, values=string.split(","))
+
+
+@dataclasses.dataclass
+class IntegerMetadataFilter(MetadataFilter):
+    low: Optional[int] = None
+    high: Optional[int] = None
+
+    class RangeModel(BaseModel):
+        from_: Optional[int] = Field(alias="from")
+        to: Optional[int]
+
+    def __post_init__(self):
+        if self.low is None and self.high is None:
+            raise ValueError("One of `low` or `high` value must be provided")
+
+    @classmethod
+    def from_string(cls, metadata_property: MetadataProperty, string: str) -> "IntegerMetadataFilter":
+        model = cls.RangeModel.parse_raw(string)
+        return cls(metadata_property, model.from_, model.to)
+
+
+@dataclasses.dataclass
+class FloatMetadataFilter(MetadataFilter):
+    low: Optional[float] = None
+    high: Optional[float] = None
+
+    class RangeModel(BaseModel):
+        from_: Optional[float] = Field(alias="from")
+        to: Optional[float]
+
+    def __post_init__(self):
+        if self.low is None and self.high is None:
+            raise ValueError("One of `low` or `high` value must be provided")
+
+    @classmethod
+    def from_string(cls, metadata_property: MetadataProperty, string: str) -> "FloatMetadataFilter":
+        model = cls.RangeModel.parse_raw(string)
+        return cls(metadata_property, model.from_, model.to)
```

Comment:
maybe we could simplify this if `{Terms,Integer,Float}MetadataFilter` were `BaseModel` directly

Reply:
but we can tackle this in another PR